language/move-lang/src/shared/unique_map.rs

```diff
@@ -0,0 +1,254 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use super::*;
+use std::{collections::BTreeMap, iter::IntoIterator};
+
+//**************************************************************************************************
+// UniqueMap
+//**************************************************************************************************
+
+/// Unique wrapper around `BTreeMap` that throws on duplicate inserts
+#[derive(Default, Clone, Debug)]
+pub struct UniqueMap<K: TName, V>(BTreeMap<K::Key, (K::Loc, V)>);
+
+impl<K: TName, V> UniqueMap<K, V> {
+    pub fn new() -> Self {
+        UniqueMap(BTreeMap::new())
+    }
+
+    pub fn is_empty(&self) -> bool {
```

Comment:
It might be worthwhile looking at the `extend` crate: https://crates.io/crates/extend

With this you could add in another method over the map that would allow you to add additional functionality in a nice way without having to write all this boiler plate.

Reply:
It looks a little different, since what I would want is auto delegates or something. This seems to be just a way of adding additional methods with constraints?

Reply:
Yea, it doesn't quite work in your case since you need to delegate all the calls (even those that overlap) so I don't think it would really work in your case. But just in case ;)